src/Collection/UniqueObjectCollection.php

```diff
@@ -0,0 +1,139 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Gamee\Collections\Collection;
+
+abstract class UniqueObjectCollection implements IUniqueObjectCollection
+{
+
+	/**
+	 * @var array
+	 */
+	private $items = [];
+
+
+	/**
+	 * Skips items with duplicate key
+	 * @param array $data
+	 */
+	public function __construct(array $data)
+	{
+		$classItemName = $this->getItemType();
+
+		\reset($data);
+		while (key($data) !== null) {
+			$item = current($data);
+
+			if (!$item instanceof $classItemName) {
+				throw new \InvalidArgumentException(static::class . ' only accepts ' . $this->getItemType());
+			}
+
+			$identifier = $this->getIdentifier($item);
+			if (!$this->exists($identifier)) {
+				$this->items[$identifier] = $item;
+			}
+
+			next($data);
+		}
+	}
+
+
+	public static function createFromImmutableObjectCollection(
+		ImmutableObjectCollection $immutableObjectCollection
+	): IUniqueObjectCollection
+	{
+		$privateAccessor = \Closure::bind(
+			static function (ImmutableObjectCollection $immutableObjectCollection) {
+				return $immutableObjectCollection->data;
+			},
+			null,
+			$immutableObjectCollection
+		);
+
+		return new static($privateAccessor($immutableObjectCollection));
+	}
+
+
+	protected function getItems(): array
+	{
+		return $this->items;
+	}
+
+
+	/**
+	 * @param mixed $item
+	 *
+	 * @return IUniqueObjectCollection
+	 * @throws DuplicateKeyException
+	 */
+	public function addItem($item): IUniqueObjectCollection
```

Comment:
Maybe there should be also the item type check as in the constructor?

Reply:
Definitely yes, added